backend/tests/test_oauth.py

```
@@ -1576,6 +1576,65 @@ def test_callback_unknown_state_redirects_with_error():
         assert "oauth=error" in r.headers["location"]
+        # A coarse reason rides along so the SPA's toast can name the failure class.
+        assert "reason=expired_or_superseded" in r.headers["location"]
+
+
+def test_callback_failure_reasons_are_self_authored_codes(monkeypatch):
+    # Each failure class carries a distinct, LITERAL reason code — never a
+    # provider-supplied string, so the redirect stays free of remote input (no open
+    # redirect / header injection) while still telling the operator what happened.
+    with TestClient(app) as c:
+
+        async def _boom(state, code):
```

Ruff (0.16.1)

[warning] Missing return type annotation for private function `_boom`

Add return type annotation: `NoReturn`

(ANN202)